util/src/main/java/tech/pegasys/artemis/util/bls/BLSSignature.java

```diff
@@ -15,50 +15,103 @@
 
 import java.util.Objects;
 import net.consensys.cava.bytes.Bytes;
+import net.consensys.cava.bytes.Bytes48;
 import net.consensys.cava.ssz.SSZ;
+import tech.pegasys.artemis.util.mikuli.BLS12381;
+import tech.pegasys.artemis.util.mikuli.KeyPair;
+import tech.pegasys.artemis.util.mikuli.PublicKey;
 import tech.pegasys.artemis.util.mikuli.Signature;
 
 public final class BLSSignature {
 
   /**
-   * Creates a random, but valid, signature
+   * Create a signature by signing the given message and domain with the given private key
    *
-   * @return the signature
+   * @param keyPair the public and private key
+   * @param message ***This will change to the digest of the message***
+   * @param domain the signature domain as per the Eth2 spec
+   * @return the resulting signature
+   */
+  public static BLSSignature sign(BLSKeyPair keyPair, Bytes message, long domain) {
+    return new BLSSignature(
+        BLS12381
+            .sign(new KeyPair(keyPair.secretKey(), keyPair.publicKey()), message, domain)
+            .signature());
+  }
+
+  /**
+   * Create a random, but valid, signature
+   *
+   * @return a random signature
    */
   public static BLSSignature random() {
     return new BLSSignature(Signature.random());
   }
 
   /**
-   * Creates an empty signature (all zero bytes)
+   * Create an empty signature (all zero bytes) as defined in the Eth2 BLS spec
+   *
+   * <p>Due to the flags, this is not actually a valid signature, so we need some extra logic to
+   * flag it as empty
    *
-   * @return the signature
+   * @return the empty signature as per the Eth2 spec
    */
   public static BLSSignature empty() {
-    return new BLSSignature(Signature.decode(Bytes.of(new byte[192])));
+    BLSSignature signature = new BLSSignature(Signature.decode(Bytes.of(new byte[192])));
+    signature.isEmpty = true;
+    return signature;
   }
 
   public static BLSSignature fromBytes(Bytes bytes) {
+    // TODO: this doesn't handle the empty signature. Do we ever need to SSZ deserialise this?
     return SSZ.decode(
         bytes, reader -> new BLSSignature(Signature.decodeCompressed(reader.readBytes())));
   }
 
-  private Signature signature;
+  private final Signature signature;
+  private boolean isEmpty = false;
 
   public BLSSignature(Signature signature) {
     this.signature = signature;
   }
 
+  /**
+   * Verify the signature against the given public key, message and domain
+   *
+   * @param publicKey the public key of the key pair that signed the message
+   * @param message the message
+   * @param domain the domain as specified in the Eth2 spec
+   * @return true if the signature is valid, false if it is not
+   */
+  boolean checkSignature(Bytes48 publicKey, Bytes message, long domain) {
+    return !isEmpty && BLS12381.verify(PublicKey.fromBytes(publicKey), signature, message, domain);
```

Comment:
I'm curious why this flag is needed.  An empty signature should be 96 bytes of 0s, right?  If you are just trying to prevent an error in BLS12381.verify(), then it would be less kludgy to have it return false if the provided point is not valid.

Reply:
`verify()` will indeed do that if passed the empty signature. The problem is that due to the flags defined in the spec, no genuine signature will serialise as 96 * zero bytes: for example, the `c` flag is always one and is always the most significant bit of the serialisation. Serialising the empty signature is required when doing [this](https://github.com/PegaSysEng/artemis/pull/378/files#diff-e59662fb0f2c311c8404594943593d76R121) for example.

Agreed that it would be better to have a special `BLSEmptySignature` class, but note that `DepositInput` sometimes needs a real signature and sometimes the empty signature, and I lost heart sorting this out. So I hacked it with a special case in the end.

I think it will do for now, but I'll add a task to revisit this before declaring BLS done.